README.md
# MusicGen Looper Interface
Gradio implementation of [Cog MusicGen Looper](https://github.com/andreasjansson/cog-musicgen-looper) by Andreas Jansson. Uses MusicGen to generate music samples then trims them into seamless loops, and can generate multiple self-consistent variations using audio to loop.

# Run on Colab
A simple colab notebook can be found [here](https://colab.research.google.com/drive/1cGaPjpETt0eBr1RPXhu3SvWoh1AaZN-I?usp=sharing) and should provide everything you need get started with the interface.

![alt text](https://github.com/sebastienlarivee/musicgen-looper-interface/blob/main/interface_screenshot.png)

# Licenses
* All code in this repository is licensed under the Apache License 2.0 license.
* The code in the [Cog MusicGen Looper](https://github.com/andreasjansson/cog-musicgen-looper) repository is released under the Apache License 2.0 as found in the [LICENSE file](https://github.com/andreasjansson/cog-musicgen-looper/blob/main/LICENSE).


CHANGELOG.txt
Version 0.2.0 (December 20 2023):
Complete code rewrite:
- All the inference code is now in 'generate.py'.
- 4 main inference functions (only 2 are implemented in Gradio in this release).
- Gradio interface now has two different "endpoints" for text to loop and audio to loop.
- New file 'globals.py' which handles model loading and save path operations.
- Code is generally much cleaner and readable (I hope!) and should speed up future development.

Interface update:
- Tabs for text to loop and audio to loop.
- Removed output file type option from 'settings', now defaults to .wav

Other changes/notes:
- Renamed 'requirements.txt' to 'requirements_colab.txt' as that's more accurate.
- Added 'CHANGELOG.txt'
- I currently don't know the full install requirements for this project as I can't run it locally.

Version 0.1.1 (December 14 2023):
I made some small changes to the initial release. I don't remeber what they were but I'll mention it in this doc for completeness.

Version 0.1.0 (December 13 2023):
Initial release.
Simple reproduction of Andreas Jansson's Cog MusicGen Looper

app.py
import gradio as gr
import globals as glo
import uuid

# from looper import main_predictor
from generate import Generate


max_audio_outputs = 10
output_folder_name = "Outputs"


def get_random_string(length=5):
    random_string = str(uuid.uuid4()).replace("-", "")[:length]
    return random_string


def model_loader(model, model_path):
    # Load custom model or base release
    if model == "custom model":
        # Need to test if download links work for this:
        glo.load_model(model_path)
    else:
        glo.load_model(f"facebook/musicgen-{model}")


def new_loops_from_text(
    model_version: str,
    custom_model_path: str,
    save_path: str,
    batch_size: int,
    bpm: int,
    text_prompt: str,
    duration: float,
    temperature: int,
    cfg_coef: float,
    seed: int,
):
    model_loader(model=model_version, model_path=custom_model_path)

    # Make the output folder(s) in the user specified location (make more efficient?)
    glo.create_output_folders(save_path, output_folder_name=output_folder_name)

    text_prompt = text_prompt + f", {bpm} bpm"
    output = []
    random_string = get_random_string()

    # Pass parameters from the gradio interface to the generation code
    predict = Generate(
        bpm=bpm,
        text_prompt=text_prompt,
        audio_prompt=None,
        duration=float(duration),
        temperature=temperature,
        cfg_coef=cfg_coef,
        seed=int(seed),
    )
    predict.set_generation_params()

    for i in range(batch_size):
        name = f"{random_string}_generation_{i+1:02d}"
        output.append(predict.loop_generate_from_text(name=name))

    # Pad with empty outputs so the returned number of outputs == max_audio_outputs
    padded_output = output + [None] * (max_audio_outputs - len(output))

    return padded_output


def new_loops_from_audio(
    model_version: str,
    custom_model_path: str,
    save_path: str,
    batch_size: int,
    bpm: int,
    text_prompt: str,
    audio_prompt: str,
    duration: float,
    temperature: int,
    cfg_coef: float,
    seed: int,
):
    model_loader(model=model_version, model_path=custom_model_path)

    # Make the output folder(s) in the user specified location (make more efficient?)
    glo.create_output_folders(save_path, output_folder_name=output_folder_name)

    text_prompt = text_prompt  # + f", {bpm} bpm"
    output = []
    random_string = get_random_string()

    # Pass parameters from the gradio interface to the generation code
    predict = Generate(
        bpm=bpm,
        text_prompt=text_prompt,
        audio_prompt=audio_prompt,
        duration=float(duration),
        temperature=temperature,
        cfg_coef=cfg_coef,
        seed=int(seed),
    )

    for i in range(batch_size):
        name = f"{random_string}_continuation_{i+1:02d}"
        output.append(predict.loop_generate_from_audio(name=name))

    # Pad with empty outputs so the returned number of outputs == max_audio_outputs
    padded_output = output + [None] * (max_audio_outputs - len(output))

    return padded_output


##################
# GRADIO INTERFACE
##################


# Handles the rendering of variable audio outputs
def variable_outputs(k):
    k = int(k)
    return [gr.Audio(type="filepath", visible=True)] * k + [
        gr.Audio(type="filepath", visible=False)
    ] * (max_audio_outputs - k)


with gr.Blocks() as interface:
    # Generate from text tab
    with gr.Tab("Text to loop"):
        with gr.Row():
            with gr.Column():
                text_prompt_gen = gr.Textbox(
                    label="Prompt",
                    placeholder="chill lofi beat, hot summer day, relaxing",
                )

                with gr.Row():
                    bpm_slider_gen = gr.Slider(
                        minimum=50, maximum=250, value=100, label="BPM"
                    )
                    duration_slider_gen = gr.Slider(
                        minimum=5, maximum=30, value=10, step=0.5, label="Max Duration"
                    )
                    batch_slider_gen = gr.Slider(
                        minimum=1,
                        maximum=max_audio_outputs,
                        value=1,
                        step=1,
                        label="Variations",
                    )

                with gr.Row():
                    seed_input_gen = gr.Textbox(value=-1, label="Seed")
                    temperature_slider_gen = gr.Slider(
                        minimum=0, maximum=1, value=1, label="Temperature"
                    )
                    cfg_slider_gen = gr.Slider(
                        minimum=0, maximum=15, step=0.5, value=3, label="CFG Scale"
                    )
                submit_button_gen = gr.Button("Submit")

            with gr.Column():
                audio_outputs_gen = []
                for i in range(max_audio_outputs):
                    a = gr.Audio(type="filepath")
                    audio_outputs_gen.append(a)
        batch_slider_gen.change(variable_outputs, batch_slider_gen, audio_outputs_gen)

    # Generate continuations from audio tab
    with gr.Tab("Loop to loop"):
        with gr.Row():
            with gr.Column():
                audio_prompt_con = gr.Audio(type="filepath")
                text_prompt_con = gr.Textbox(
                    label="Prompt",
                    placeholder="chill lofi beat, hot summer day, relaxing",
                )

                with gr.Row():
                    bpm_slider_con = gr.Slider(
                        minimum=50, maximum=250, value=100, label="BPM"
                    )
                    duration_slider_con = gr.Slider(
                        minimum=5, maximum=30, value=10, step=0.5, label="Max Duration"
                    )
                    batch_slider_con = gr.Slider(
                        minimum=1,
                        maximum=max_audio_outputs,
                        value=1,
                        step=1,
                        label="Variations",
                    )

                with gr.Row():
                    seed_input_con = gr.Textbox(value=-1, label="Seed")
                    temperature_slider_con = gr.Slider(
                        minimum=0, maximum=1, value=0.95, label="Temperature"
                    )
                    cfg_slider_con = gr.Slider(
                        minimum=0, maximum=15, step=0.5, value=3, label="CFG Scale"
                    )

                submit_button_con = gr.Button("Submit")
            with gr.Column():
                audio_outputs_con = []
                for i in range(max_audio_outputs):
                    a = gr.Audio(type="filepath")
                    audio_outputs_con.append(a)
        batch_slider_con.change(variable_outputs, batch_slider_con, audio_outputs_con)

    # Settings tab
    with gr.Tab("Settings"):
        with gr.Column():
            with gr.Row():
                model_toggle_set = gr.Radio(
                    choices=[
                        "small",
                        "medium",
                        "large",
                        "stereo-small",
                        "stereo-medium",
                        "stereo-large",
                        "custom model",
                    ],
                    value="stereo-medium",
                    label="Model Version",
                )
                model_path_set = gr.Textbox(
                    label="Custom Model Path",
                    placeholder="File path to your model",
                )
        with gr.Column():
            with gr.Row():
                save_path_set = gr.Textbox(
                    value="/content/musicgen-looper-interface", label="Save Path"
                )
    # Action handlers (NEED TO CLEAN THESE UP FOR NEXT RELEASE)
    submit_button_gen.click(
        fn=new_loops_from_text,
        inputs=[
            model_toggle_set,
            model_path_set,
            save_path_set,
            batch_slider_gen,
            bpm_slider_gen,
            text_prompt_gen,
            duration_slider_gen,
            temperature_slider_gen,
            cfg_slider_gen,
            seed_input_gen,
        ],
        outputs=audio_outputs_gen,
    )
    submit_button_con.click(
        fn=new_loops_from_audio,
        inputs=[
            model_toggle_set,
            model_path_set,
            save_path_set,
            batch_slider_con,
            bpm_slider_con,
            text_prompt_con,
            audio_prompt_con,
            duration_slider_con,
            temperature_slider_con,
            cfg_slider_con,
            seed_input_con,
        ],
        outputs=audio_outputs_con,
    )


if __name__ == "__main__":
    print("Launching MusicGen Looper interface...")
    interface.launch(share=True)


to-do.py
import that_dawg

# ITS TIME TO REFACTOR

# Gonna split all of this into seperate processes:

### Model loading DONE
### Folder creation DONE
### Basic inference DONE
### Beat counting + trimming DONE
### Beat synced continuation
### Bar counting logic
### Changelog file
### Folders based on generation type

# Flow thoughts:

# Generation tab generates individual loops
### The loops are beat synced and sliced
### Can send audio made here to continuation, assembly

# Continuations tab generates beat synced continuations
### Meant to receive input from generations tab
### Continuations can be fed back into the continuations tab to make more
### Can send audio made here to continuation, assembly

# Assembly tab (probably not in scope for this refactor)
### Collects audio files sent to it from other tabs
### Audio here can be rearanged, looped, and played back
### Basically a mini-daw that's gonna be pretty janky
### Can export the full length final assembly


globals.py
import torch
from audiocraft.models import MusicGen
import os
import datetime


# Store all global variable which are passed to most functions
MODEL = None
INTERRUPTED = False
SAVE_PATH = ""


def interrupt():
    global INTERRUPTED
    INTERRUPTED = True
    print("Interrupted!")


# Loading function from the fb gradio demo
def load_model(version):
    global MODEL
    print("Loading model", version)
    if MODEL is None or MODEL.name != version:
        if MODEL is not None:
            # Clear PyTorch CUDA cache and delete model
            del MODEL
            torch.cuda.empty_cache()
        MODEL = None  # in case loading would crash
        MODEL = MusicGen.get_pretrained(version)


# Create save path for generations
def create_output_folders(base_path, output_folder_name):
    global SAVE_PATH
    # Ensure base_path ends with a slash
    if not base_path.endswith(os.sep):
        base_path += os.sep

    # Create 'Outputs' folder
    outputs_path = base_path + output_folder_name
    if not os.path.exists(outputs_path):
        os.makedirs(outputs_path)

    # Create yyyy-mm-dd folder
    current_date = datetime.date.today().isoformat()
    date_folder_path = os.path.join(outputs_path, current_date)
    date_folder_path += os.sep
    if not os.path.exists(date_folder_path):
        os.makedirs(date_folder_path)

    # Return the path to the folder
    SAVE_PATH = date_folder_path


generate.py
import globals as glo
import os
import random
import torch
import torchaudio
import numpy as np
import soundfile as sf
import librosa
import pyrubberband as pyrb
from BeatNet.BeatNet import BeatNet
import madmom.audio.filters

# Hack madmom to work with recent python (need to figure out that madmom is actually doing)
madmom.audio.filters.np.float = float


class Generate:
    def __init__(
        self,
        bpm: int,
        text_prompt: str,
        audio_prompt: str,
        duration: float,
        temperature: int,
        cfg_coef: float,
        seed: int,
    ):
        # Global variables:
        self.save_path = glo.SAVE_PATH
        self.model = glo.MODEL
        self.model_sample_rate = glo.MODEL.sample_rate

        # Variables passed from app.py
        self.bpm = bpm
        self.text_prompt = text_prompt

        if audio_prompt is not None:
            self.audio_prompt, self.prompt_sample_rate = torchaudio.load(audio_prompt)
            self.audio_prompt_wav = audio_prompt

        self.duration = duration
        self.temperature = temperature
        self.cfg_coef = cfg_coef

        # Gets a random seed if none is specified
        if not seed or seed == -1:
            self.seed = torch.seed() % 2**32 - 1
        else:
            self.seed = seed

    def set_generation_params(self):
        # Updates the MusicGen model parameters to user input
        self.model.set_generation_params(
            duration=self.duration,
            top_k=250,
            top_p=0,
            temperature=self.temperature,
            cfg_coef=self.cfg_coef,
        )

    def set_all_seeds(self):
        # Sets seed for the generation
        random.seed(self.seed)
        os.environ["PYTHONHASHSEED"] = str(self.seed)
        np.random.seed(self.seed)
        torch.manual_seed(self.seed)
        torch.cuda.manual_seed(self.seed)
        torch.backends.cudnn.deterministic = True

    def estimate_beats(self, wav):
        # Maps out the beats
        beatnet = BeatNet(
            1,
            mode="offline",
            inference_model="DBN",
            plot=[],
            thread=False,
            device="cuda:0",
        )

        beatnet_input = librosa.resample(
            wav, orig_sr=self.model_sample_rate, target_sr=beatnet.sample_rate
        )
        return beatnet.process(beatnet_input)

    def get_loop_points(self, beats, wav):
        # Trims + stretches the audio to match user BPM as a seamless loop
        downbeat_times = beats[:, 0][beats[:, 1] == 1]
        num_bars = len(downbeat_times) - 1
        if num_bars < 1:
            raise ValueError(
                "Less than one bar detected. Try increasing max_duration, or use a different seed."
            )
        even_num_bars = int(2 ** np.floor(np.log2(num_bars)))
        start_time = downbeat_times[0]
        end_time = downbeat_times[even_num_bars]
        num_beats = len(beats[(beats[:, 0] >= start_time) & (beats[:, 0] < end_time)])
        duration = end_time - start_time
        actual_bpm = num_beats / duration * 60

        # Handle possible octave errors
        if abs(actual_bpm / 2 - self.bpm) <= 10:
            actual_bpm = actual_bpm / 2
        elif abs(actual_bpm * 2 - self.bpm) <= 10:
            actual_bpm = actual_bpm * 2

        # Prepare the main audio loop
        start_sample = int(start_time * self.model_sample_rate)
        end_sample = int(end_time * self.model_sample_rate)
        loop = wav[start_sample:end_sample]

        # Process the audio loop for the main output
        stretched = pyrb.time_stretch(
            loop, self.model_sample_rate, self.bpm / actual_bpm
        )
        return stretched

    def write(self, audio, name):
        # Save's as file type given by user
        wav_path = self.save_path + name + ".wav"
        print(wav_path)
        sf.write(wav_path, audio, self.model_sample_rate)
        return wav_path

    def generate_from_text(self):
        # Generates audio from a text prompt
        self.set_all_seeds()

        generation = (
            self.model.generate([self.text_prompt], progress=True).cpu().numpy()[0, 0]
        )
        generation = generation / np.abs(generation).max()
        return generation

    def generate_from_audio(self, prompt=None, start_indice=None, end_indice=None):
        # Generates audio from an audio prompt
        if prompt == None:
            prompt = self.audio_prompt

        self.set_all_seeds()
        generation = (
            self.model.generate_continuation(
                prompt=prompt,
                prompt_sample_rate=self.prompt_sample_rate,
                descriptions=[self.text_prompt],
                progress=True,
            )
            .cpu()
            .numpy()[0, 0, start_indice:end_indice]
            # ^ Optional parameter to slice off the input audio
        )
        return generation

    ################################################
    # ENDPOINTS TO CALL THROUGH THE GRADIO INTERFACE
    ################################################

    def simple_generate_from_text(self, name):
        # ENDPOINT: Generate audio from a text prompt
        wav = self.predict_from_text()
        output_path = self.write(audio=wav, name=name)
        self.seed += 1  # For batch generation
        return output_path

    def simple_generate_from_audio(self, name):
        # ENDPOINT: Generate audio from an audio prompt
        prompt_duration = 3  # Placeholder value for testing

        self.audio_prompt = self.audio_prompt[
            ..., -int(prompt_duration * self.prompt_sample_rate) :
        ]

        wav = self.generate_from_audio(prompt=self.audio_prompt)
        output_path = self.write(audio=wav, name=name)
        self.seed += 1  # For batch generation
        return output_path

    def loop_generate_from_text(self, name):
        # ENDPOINT: Generate seamless loops from a text prompt
        print(f"Text to loop, seed: {self.seed}")
        wav = self.generate_from_text()
        beats = self.estimate_beats(wav=wav)
        loop = self.get_loop_points(beats=beats, wav=wav)
        output_path = self.write(audio=loop, name=name)
        self.seed += 1  # For batch generation
        return output_path

    def loop_generate_from_audio(self, name):
        # ENDPOINT: Generate seamless loops from an audio prompt (prompt must also be a loop to work well)
        print(f"Audio to loop, seed: {self.seed}")

        original_loop_seconds = self.audio_prompt.size(1) / self.prompt_sample_rate

        # Might as well give full context if we can?
        if original_loop_seconds * 2 <= 29.8:
            prompt_seconds = original_loop_seconds
            prompt = self.audio_prompt
        else:
            prompt_seconds = 29.8 - original_loop_seconds
            prompt = self.audio_prompt[
                ..., -int(prompt_seconds * self.prompt_sample_rate) :
            ]

        # Set total duration (+ a bit of extra) and update generation params
        self.duration = prompt_seconds + original_loop_seconds + 0.1
        self.set_generation_params()

        # Slice generated audio from the end of the prompt to the length of the original loop
        # This needs to be calculated with model_sample_rate
        start_indice = int(prompt_seconds * self.model_sample_rate)
        end_indice = int(start_indice + original_loop_seconds * self.model_sample_rate)

        # Generate
        wav = self.generate_from_audio(
            prompt=prompt,
            start_indice=start_indice,
            end_indice=end_indice,
        )

        # Blending (I feel like this can be improved)
        num_lead = 100
        lead = sf.read(self.audio_prompt_wav, start=-num_lead)
        lead = lead[0]
        num_lead = len(lead)
        wav[-num_lead:] *= np.linspace(1, 0, num_lead)
        wav[-num_lead:] += np.linspace(0, 1, num_lead) * lead

        output_path = self.write(audio=wav, name=name)
        self.seed += 1  # For batch generation
        return output_path


looper.py
# This code is adapted from:
# https://github.com/andreasjansson/cog-musicgen-looper by Andreas Jansson (Apache 2.0 license)
# With modifications

# You'll need to run this line for colab and stuff:
# !apt-get install -y ffmpeg portaudio19-dev rubberband-cli


import os
import datetime
import random
import uuid
import gc
import subprocess
import torch
import numpy as np
import soundfile as sf
import librosa
import pyrubberband as pyrb
from audiocraft.models import MusicGen
from audiocraft.models.loaders import (
    load_compression_model,
    load_lm_model,
)
from BeatNet.BeatNet import BeatNet
import madmom.audio.filters

# Hack madmom to work with recent python
madmom.audio.filters.np.float = float

loaded_models = {}
output_folder_name = "Outputs"


def set_all_seeds(seed):
    random.seed(seed)
    os.environ["PYTHONHASHSEED"] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True


def generate_random_string(length=5):
    random_str = str(uuid.uuid4()).replace("-", "")[:length]
    return random_str


def estimate_beats(wav, sample_rate, beatnet):
    beatnet_input = librosa.resample(
        wav, orig_sr=sample_rate, target_sr=beatnet.sample_rate
    )
    return beatnet.process(beatnet_input)


def get_loop_points(beats):
    downbeat_times = beats[:, 0][beats[:, 1] == 1]
    num_bars = len(downbeat_times) - 1
    if num_bars < 1:
        raise ValueError(
            "Less than one bar detected. Try increasing max_duration, or use a different seed."
        )
    even_num_bars = int(2 ** np.floor(np.log2(num_bars)))
    start_time = downbeat_times[0]
    end_time = downbeat_times[even_num_bars]
    return start_time, end_time


def write(audio, sample_rate, output_format, name):
    wav_path = name + ".wav"
    sf.write(wav_path, audio, sample_rate)
    if output_format == "mp3":
        mp3_path = name + ".mp3"
        subprocess.call(
            ["ffmpeg", "-loglevel", "error", "-y", "-i", wav_path, mp3_path]
        )
        os.remove(wav_path)
        path = mp3_path
    else:
        path = wav_path
    return path


def create_output_folders(base_path):
    # Ensure base_path ends with a slash
    if not base_path.endswith(os.sep):
        base_path += os.sep

    # Create 'Outputs' folder
    outputs_path = base_path + output_folder_name
    if not os.path.exists(outputs_path):
        os.makedirs(outputs_path)

    # Create yyyy-mm-dd folder
    current_date = datetime.date.today().isoformat()
    date_folder_path = os.path.join(outputs_path, current_date)
    if not os.path.exists(date_folder_path):
        os.makedirs(date_folder_path)

    # Return the path to the folder
    return date_folder_path


def load_fb_model(model_id, device):
    # Check if the model is already loaded
    if model_id in loaded_models:
        return loaded_models[model_id]

    # Load the model since it's not loaded yet
    print(f"Loading model: {model_id}...")
    compression_model = load_compression_model(model_id, device=device)
    lm = load_lm_model(model_id, device=device)
    musicgen_model = MusicGen(model_id, compression_model, lm)

    # Store the loaded model in the dictionary
    loaded_models[model_id] = musicgen_model
    print(f"Model loaded!")
    return musicgen_model


def load_custom_model(custom_model_path):
    # Check if the model is already loaded
    if custom_model_path in loaded_models:
        return loaded_models[custom_model_path]

    # Load the model since it's not loaded yet
    print("Loading custom model...")
    musicgen_model = MusicGen.get_pretrained(custom_model_path)

    # Store the loaded model in the dictionary
    loaded_models[custom_model_path] = musicgen_model
    print("Model loaded!")
    return musicgen_model


def main_predictor(params):
    device = "cuda" if torch.cuda.is_available() else "cpu"
    # Determine which model to load based on the 'model_version' parameter
    model_version = params["model_version"]

    if model_version == "custom model":
        custom_model_path = params["custom_model_path"]
        model_id = custom_model_path
        model = load_custom_model(custom_model_path)
    else:
        model_id = f"facebook/musicgen-{model_version}"  # Matches audiocraft naming scheme, needs to be more robust
        model = load_fb_model(model_id, device)

    beatnet = BeatNet(
        1, mode="offline", inference_model="DBN", plot=[], thread=False, device="cuda:0"
    )

    bpm = params["bpm"]
    seed = params["seed"]
    top_k = params["top_k"]
    top_p = params["top_p"]
    prompt = params["prompt"]
    variations = params["variations"]
    temperature = params["temperature"]
    max_duration = params["max_duration"]
    model_version = params["model_version"]
    output_format = params["output_format"]
    guidance = params["classifier_free_guidance"]
    base_save_path = params["save_path"]

    save_path = create_output_folders(base_save_path)

    model.set_generation_params(
        duration=max_duration,
        top_k=top_k,
        top_p=top_p,
        temperature=temperature,
        cfg_coef=guidance,
    )

    if not seed or seed == -1:
        seed = torch.seed() % 2**32 - 1
    set_all_seeds(seed)

    print(
        f"Generating: {model_version}, {variations} variation(s), prompt: {prompt}, seed: {seed}"
    )

    prompt = prompt + f", {bpm} bpm"
    print("Variation 01: generating...")
    wav = model.generate([prompt], progress=True).cpu().numpy()[0, 0]
    wav = wav / np.abs(wav).max()

    beats = estimate_beats(wav, model.sample_rate, beatnet)
    start_time, end_time = get_loop_points(beats)
    num_beats = len(beats[(beats[:, 0] >= start_time) & (beats[:, 0] < end_time)])
    duration = end_time - start_time
    actual_bpm = num_beats / duration * 60

    # Handle possible octave errors
    if abs(actual_bpm / 2 - bpm) <= 10:
        actual_bpm = actual_bpm / 2
    elif abs(actual_bpm * 2 - bpm) <= 10:
        actual_bpm = actual_bpm * 2

    # Prepare the main audio loop
    start_sample = int(start_time * model.sample_rate)
    end_sample = int(end_time * model.sample_rate)
    loop = wav[start_sample:end_sample]

    # Process the audio loop for the main output
    stretched = pyrb.time_stretch(loop, model.sample_rate, bpm / actual_bpm)

    # Generate a random string for this set of variations
    random_string = generate_random_string()

    # Initialize outputs list
    outputs = []

    # Save the main output
    main_output_path = write(
        stretched,
        model.sample_rate,
        output_format,
        f"{save_path}variation_01_{random_string}",
    )
    outputs.append(main_output_path)  # Append to list instead of dictionary
    # print(f"Outputs list: {outputs}")

    # Generate additional variations if requested
    if variations > 1:
        # Use last 4 beats as audio prompt
        last_4beats = beats[beats[:, 0] <= end_time][-5:]
        audio_prompt_start_time = last_4beats[0][0]
        audio_prompt_end_time = last_4beats[-1][0]
        audio_prompt_start_sample = int(audio_prompt_start_time * model.sample_rate)
        audio_prompt_end_sample = int(audio_prompt_end_time * model.sample_rate)
        audio_prompt_seconds = audio_prompt_end_time - audio_prompt_start_time
        audio_prompt = torch.tensor(
            wav[audio_prompt_start_sample:audio_prompt_end_sample]
        )[None]
        audio_prompt_duration = audio_prompt_end_sample - audio_prompt_start_sample

        model.set_generation_params(
            duration=duration + audio_prompt_seconds + 0.1,
            top_k=top_k,
            top_p=top_p,
            temperature=temperature,
            cfg_coef=guidance,
        )

        for i in range(2, variations + 1):
            print(f"Variation {i:02d}: generating...")
            continuation = model.generate_continuation(
                prompt=audio_prompt,
                prompt_sample_rate=model.sample_rate,
                descriptions=[prompt],
                progress=True,
            )
            variation_loop = continuation.cpu().numpy()[
                0, 0, audio_prompt_duration : audio_prompt_duration + len(loop)
            ]

            # Process each variation loop
            num_lead = 100  # for blending to avoid clicks
            lead_start = start_sample - num_lead
            lead = wav[lead_start:start_sample]
            num_lead = len(lead)
            variation_loop[-num_lead:] *= np.linspace(1, 0, num_lead)
            variation_loop[-num_lead:] += np.linspace(0, 1, num_lead) * lead

            variation_stretched = pyrb.time_stretch(
                variation_loop, model.sample_rate, bpm / actual_bpm
            )

            # Save each variation
            variation_output_path = write(
                variation_stretched,
                model.sample_rate,
                output_format,
                f"/{save_path}variation_{i:02d}_{random_string}",
            )
            outputs.append(variation_output_path)  # Append to list
            # print(f"Outputs list: {outputs}")

    torch.cuda.empty_cache()
    gc.collect()

    # print(f"Inference outputs: {outputs}")
    return outputs
